main.go

```diff
@@ -124,118 +132,174 @@ func main() {
 		fmt.Printf("Please specify host, port, user and database using PGHOST, PGPORT, PGUSER and PGDATABASE environment variables\n")
-		return fmt.Errorf("Could not connect to source database")
+		return "", fmt.Errorf("Could not connect to source database")
 	}
 	password := getPassword()
-	sourceDB, err := sql.Open(driver, fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable", server, port, user, password, dbname))
+	return fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable", server, port, user, password, dbname), nil
+}
+
+func schemaConvFromSourceDB(driver string) (*internal.Conv, error) {
+	fmt.Println(`###################################################################
+	Accessing a source DB via an database/sql driver is an experimental
+	feature that is not fully implemented. Currently we only do schema
+	conversion and we only support PostgreSQL.
+	###################################################################`)
+
+	driverConfig, err := driverConfig(driver)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	sourceDB, err := sql.Open(driver, driverConfig)
+	if err != nil {
+		return nil, err
 	}
 	conv := internal.MakeConv()
 	err = internal.ProcessInfoSchema(conv, sourceDB)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	writeSchemaFile(conv, now, outputFilePrefix+schemaFile)
-	spannerDB, err := createDatabase(projectID, instanceID, dbName, conv)
-	if err != nil {
-		fmt.Printf("\nCan't create database: %v\n", err)
-		return fmt.Errorf("can't create database")
+	return conv, nil
+}
+
+func dataConvFromSourceDB(ioHelper *ioStreams, client *sp.Client, conv *internal.Conv) (*spanner.BatchWriter, error) {
+	config := spanner.BatchWriterConfig{
+		BytesLimit: 100 * 1000 * 1000,
+		WriteLimit: 40,
+		RetryLimit: 1000,
+		Verbose:    internal.Verbose(),
+		Write: func(m []*sp.Mutation) error {
+			return nil
+		},
 	}
-	badWrites := make(map[string]int64) // Empty bad writes since no data conversion yet.
-	banner := getBanner(now, spannerDB)
-	report(badWrites, 0, banner, conv, outputFilePrefix+reportFile)
-	return nil
+	return spanner.NewBatchWriter(config), nil
 }
 
 type ioStreams struct {
-	in, out *os.File
+	in, seekableIn, out *os.File
+	bytesRead           int64
 }
 
-var ioHelper = &ioStreams{os.Stdin, os.Stdout}
-
-func pgDumpToSpanner(projectID, instanceID, dbName string, helper *ioStreams, outputFilePrefix string, now time.Time) error {
-	if helper != nil {
-		ioHelper = helper
-	}
-
+func schemaConvFromFileStream(ioHelper *ioStreams) (*internal.Conv, error) {
 	f, n, err := getSeekable(ioHelper.in)
 	if err != nil {
-		printSeekError(err)
-		return fmt.Errorf("can't get seekable input file")
+		printSeekError(err, ioHelper.out)
+		return nil, fmt.Errorf("can't get seekable input file")
 	}
-	defer f.Close()
+	ioHelper.seekableIn = f
+	ioHelper.bytesRead = n
 	conv := internal.MakeConv()
 
 	err = firstPass(f, n, conv)
```

Comment:
How about just inlining firstPass?
The concept of 'first pass' has outlived its usefulness.
And the split between these two functions is pretty arbitrary and both functions are small.